src/acq/channel.c

```diff
@@ -365,6 +365,8 @@ void bl_acq_channel_commit_sample(enum bl_acq_source source, uint32_t sample)
 		msg->sample_data.data32[count] = sample;
 
 		if (msg->sample_data.count >= MSG_SAMPLE_DATA32_MAX) {
+			msg->sample_data.src_led = bl_led_src[bl_led_active].led;
```

Comment:
I don't understand what's happening here, is this a LED index rather than mask? If so, how is the accumulated message not composed of interleaved samples?

Reply:
This is LED index.

In bl_acq_adc_dma_isr, we accumulated samples from all channels, but only commit the ones matches active source mask, which is either the active LED or non-LED sources(Vcc and temp). This guaranteed the src_led filed in the sample data for a LED source is valid, and that for a non-LED source can just be ignored.

Reply:
This reminder me a bug in bl_acq_adc_dma_isr, where led loop should only been called once if there are non-led channels in the same adc.

Reply:
Ok, so we always send a packet per flash?

Reply:
depends on if there are non-led source enabled on the same adc. if not, there will be 1 led packet sent, then led switch, if yes, then the led + non-led channels will all been sent

Reply:
Ok, so concrete example. Let's say I flash LD13, LD14, LD15 and LD16. I'm assuming the samples are interleaved. Do we send a packet each time we change which LED is flashing or do we send mixed data? If we're sending mixed data, the field isn't an index right?

Reply:
We only send the packet belongs to currently flashing LED. So it's not sending mixed data.

For example, if the LED13 is on at moment, and other 3 LEDs are off. We sample until DMA interrupted.
line 654 in the bl_acq_adc_dma_isr will filter out the samples and only commit LED13's. Once this is done, led_loop is called, and LED14 turns on with LED13 turns off.

Reply:
ah, ok, so we do a full 15 or 31 samples of each led and then flash to another one? I guess that makes the interleaving quite a bit slower because we're taking multiple samples in a burst.

Reply:
No, we change the LED every valid sample of the LED, and the sample_data.count for that LED will only +1 each time. As there is only 1 LED be allowed to flash at a time, it's guaranteed when the count reaches 15 or 31, the flashing LED index matches the sample data.

But yes, it will be much slower, inversely proportional to the number of enabled LEDs, as for the same sample rate, the hardware are collecting the same amount of data, but we only adopt one channel for the flashing LED and drop data from all other channels.

This is ideally to be optimised, but unless we want to change adc channel config each time we switch LEDs, I don't see an easy way to do it.

Reply:
I think discarding ADC input when the relevant LED is off is the correct choice, that's fine. What I'm more trying to get to the bottom of is how the interleaving is working, ideally we'd interleave the LEDs such that the LED comes on for 1 sample period (including sw_oversample) if possible.

I fully expect the max sample rate to be lower for multiple LED's, that's not an issue at all. I'm just trying to get my head around how the interleaving works and how it should work.

I'm trying to get my head around how this works, not sure which of the following represents what a message will look like:
1. <led=0><LD1 Sample><LD1 Sample><LD1 Sample><LD1 Sample>
2. <led=0><LD1 Sample><LD2 Sample><LD3 Sample><LD4 Sample>

If the answer is 1 then the results will be bursts which are going to lead to artifacts and aren't as clean as proper interleaving. If the answer is 2 then the led field is a bit confusing to me.

I had some ideas about changing how we consider a "channel" so that each LED+PD combo gets a message queue and we have an arbitrary maximum. That way we could differentiate the messages more clearly maybe.

Reply:
I got what you mean now, and I think you are correct, the way it works is problematic.

As multiple LEDs are going to share the same PD channel, it will be pattern 2 in your example.

What's your idea about differentiate LED+PD combo?